core/src/main/java/org/apache/iceberg/jdbc/JdbcCatalog.java

```diff
@@ -0,0 +1,391 @@
+import org.apache.iceberg.catalog.SupportsNamespaces;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.exceptions.AlreadyExistsException;
+import org.apache.iceberg.exceptions.NamespaceNotEmptyException;
+import org.apache.iceberg.exceptions.NoSuchNamespaceException;
+import org.apache.iceberg.exceptions.NoSuchTableException;
+import org.apache.iceberg.exceptions.UncheckedSQLException;
+import org.apache.iceberg.hadoop.HadoopFileIO;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.relocated.com.google.common.base.Joiner;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class JdbcCatalog extends BaseMetastoreCatalog implements Configurable, SupportsNamespaces, Closeable {
+
+  public static final String SQL_TABLE_NAME = "iceberg_tables";
+  public static final String SQL_TABLE_DDL =
+          "CREATE TABLE " + SQL_TABLE_NAME +
+                  "(catalog_name VARCHAR(1255) NOT NULL," +
+                  "table_namespace VARCHAR(1255) NOT NULL," +
+                  "table_name VARCHAR(1255) NOT NULL," +
+                  "metadata_location VARCHAR(32768)," +
+                  "previous_metadata_location VARCHAR(32768)," +
+                  "PRIMARY KEY (catalog_name, table_namespace, table_name)" +
+                  ")";
+  public static final String SQL_SELECT_TABLE = "SELECT * FROM " + SQL_TABLE_NAME +
+          " WHERE catalog_name = ? AND table_namespace = ? AND table_name = ? ";
+  public static final String SQL_SELECT_ALL = "SELECT * FROM " + SQL_TABLE_NAME +
+          " WHERE catalog_name = ? AND table_namespace = ?";
+  public static final String SQL_UPDATE_TABLE_NAME = "UPDATE " + SQL_TABLE_NAME +
+          " SET table_namespace = ? , table_name = ? " +
+          " WHERE catalog_name = ? AND table_namespace = ? AND table_name = ? ";
+  public static final String SQL_DELETE_TABLE = "DELETE FROM " + SQL_TABLE_NAME +
+          " WHERE catalog_name = ? AND table_namespace = ? AND table_name = ? ";
+  public static final String SQL_SELECT_NAMESPACE = "SELECT table_namespace FROM " + SQL_TABLE_NAME +
+          " WHERE catalog_name = ? AND table_namespace LIKE ? LIMIT 1";
+  public static final String SQL_SELECT_NAMESPACES = "SELECT DISTINCT table_namespace FROM " + SQL_TABLE_NAME +
+          " WHERE catalog_name = ? AND table_namespace LIKE ?";
+  public static final String JDBC_PARAM_PREFIX = "connection.parameter.";
+  private static final Logger LOG = LoggerFactory.getLogger(JdbcCatalog.class);
+  private static final Joiner SLASH = Joiner.on("/");
+
+  private FileIO fileIO;
+  private String catalogName = "jdbc";
+  private String warehouseLocation;
+  private Configuration hadoopConf;
+  private JdbcClientPool dbConnPool;
```

Comment:
Style: in other places, we use different naming conventions:

* `FileIO` instances are typically called `io`
* `Configuration` instances are typically called `conf`
* A pool would typically be a plural named for the pooled objects, like `connections`